### ISSUE TITLE
lambda arguments don't seem to be checked

### ISSUE THREAD

**Issue by aspin** [facebook/pyre-check#295]
Not sure if there's a limitation here since it's not possible to annotate lambda arguments, but the following code sample doesn't raise any errors:

```python
class Struct(NamedTuple):
    nums: List[int]
    strs: List[str]


def process(s: Struct) -> List[int]:
    return (
        list(map(lambda x: x.real, s.nums))
        + list(map(lambda x: x.real, s.strs))
    )


if __name__ == "__main__":
    res = process(
        Struct(
            [1,2,3,4],
            ["1","2","3", "4"]
        )
    )
    print(res)
```

Expected behavior: some error is raised since `.real` doesn't exist on `str`s

**Comment by arthaud** [comment]
Hi @aspin, thanks for reaching out.

This is a known limitation of Pyre. Since lambda arguments cannot be annotated, they are implicitly typed as `Any`.

I have actually done some work on lambda support, in order to properly infer the return type using the parameter type inferred from the context (here, the second argument to `map`). We could go further and properly type check the lambda body using the inferred argument type. However, this special casing is hard to maintain and causes a recursion in the analyzer, and we try to avoid that. In normal cases, each parameter expression can be typed independently (e.g, in `f(a, b)`, typing `a` should not require information on `b`).

**Comment by aspin** [comment]
Gotcha, figured it'd be something like that.
